### PR TITLE
Condition-Space Interpolation: same-geometry augmentation

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Create training augmentations by interpolating between CFD solutions at **different conditions (AoA, Re) but same geometry**. Unlike general sample mixup (#2327 closed, catastrophic because different geometries are mixed), this only interpolates samples that share the same airfoil shape — where linear interpolation of the flow field is physically meaningful for small condition deltas.

For two samples with identical geometry at (AoA=5, Re=1e6) and (AoA=7, Re=1e6), the flow at (AoA=6, Re=1e6) is approximately the average of the two. This teaches the model that the condition manifold is smooth — directly targeting p_oodc and p_re.

### Step 1: Index same-geometry pairs in the dataset

During dataset loading, group samples by geometry ID (or approximate by DSDF-1 cosine similarity > 0.98). For each training batch, identify pairs (i, j) where:
- Same geometry (same airfoil shape)
- Different conditions (different AoA or Re)

```python
def find_same_geom_pairs(batch_features, dsdf_dim_start, dsdf_dim_end, threshold=0.98):
    """Find pairs with same geometry (high DSDF cosine similarity)."""
    dsdf = batch_features[:, 0, dsdf_dim_start:dsdf_dim_end]  # [B, D_dsdf] — take any node's DSDF
    dsdf_norm = dsdf / (dsdf.norm(dim=-1, keepdim=True) + 1e-8)
    sim = dsdf_norm @ dsdf_norm.T  # [B, B]
    pairs = (sim > threshold).nonzero()  # same-geometry pairs
    # Filter: different conditions (AoA differs by > 0.5 degree)
    aoa = batch_features[:, 0, 14]  # AoA feature
    valid = (aoa[pairs[:, 0]] - aoa[pairs[:, 1]]).abs() > 0.01
    return pairs[valid]
```

### Step 2: Interpolate same-geometry pairs

```python
if args.condition_interp and len(pairs) > 0:
    # Pick up to 4 pairs per batch
    idx = torch.randperm(len(pairs))[:4]
    selected = pairs[idx]
    
    lam = torch.rand(len(selected), 1, 1, device=x.device) * 0.4 + 0.3  # [0.3, 0.7]
    
    x_interp = lam * x[selected[:, 0]] + (1-lam) * x[selected[:, 1]]
    y_interp = lam * y[selected[:, 0]] + (1-lam) * y[selected[:, 1]]
    
    # Compute loss on interpolated samples
    y_pred_interp = model(x_interp)
    loss_interp = F.l1_loss(y_pred_interp, y_interp)
    
    total_loss = total_loss + args.condition_interp_weight * loss_interp
```

### Step 3: Flags

- `--condition_interp` (bool)
- `--condition_interp_weight` (float, default=0.3)

### Key constraints
- Only interpolate same-geometry pairs (DSDF cosine sim > 0.98)
- Only from in_dist and ood_cond sources (NOT tandem, where geometry changes)
- Lambda in [0.3, 0.7] (avoid extremes)
- Limit to 4 pairs per batch to avoid slowing training

### Training commands (2 seeds)

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent edward --wandb_name "edward/cond-interp-s42" --wandb_group condition-interp-augmentation \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --condition_interp --condition_interp_weight 0.3

CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent edward --wandb_name "edward/cond-interp-s73" --wandb_group condition-interp-augmentation \
  --seed 73 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --condition_interp --condition_interp_weight 0.3
```

## Baseline
| Metric | Value | Target |
|--------|-------|--------|
| p_in | 11.709 | < 11.709 |
| p_oodc | 7.544 | < 7.544 |
| p_tan | 27.402 | < 27.402 |
| p_re | 6.481 | < 6.481 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Research & Implementation Note

**Physical validity**: Checked with researcher-agent. Linear interpolation of RANS flow fields is physically valid for small AoA deltas (< 2-3°) in attached flow. In the linear aerodynamic regime, pressure distributions change quasi-linearly with AoA. The interpolated labels are approximate pseudo-solutions rather than exact RANS solutions, but provide a physically grounded smoothness constraint.

**Key implementation decisions**:
1. **Batch concatenation**: Interpolated samples are appended to the original batch and processed in one forward pass (efficient). `condition_interp_weight` is applied via per-sample weighting of `abs_err` before loss computation.
2. **Non-tandem only**: Filtered to single-foil samples only (`gap=0`). Same-geometry tandem pairs would require matching both gap AND stagger — too restrictive to find pairs.
3. **DSDF cosine similarity** computed on `x[:, 0, 2:10]` (first node's 8-dim DSDF) — a proxy for geometry identity. Threshold 0.98.
4. **AoA filter**: `|AoA_i - AoA_j| > 0.01 rad` (≈ 0.57°) to avoid near-identical conditions.
5. **Lambda [0.3, 0.7]**: Avoids degenerate near-copies, stays in the meaningful interpolation regime.

W&B runs started:
- seed 42: `bota4ckq`
- seed 73: `65eo5fnu`

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Seed 42 — W&B run `bota4ckq`
- **Best epoch:** 136 (wall-clock timeout at 180.5 min)
- **val/loss:** 0.6246
- **Surface MAE (p):**
  - p_in: 13.3 (baseline 11.709, **+13.6% worse**)
  - p_tan: 27.5 (baseline 27.402, **+0.4% worse**)
  - p_oodc: 7.8 (baseline 7.544, **+3.4% worse**)
  - p_re: 6.9 (baseline 6.481, **+6.5% worse**)
- **Peak VRAM:** 67.8 GB (vs ~40 GB baseline — significant overhead from interp samples)
- **Command:**
  ```
  CUDA_VISIBLE_DEVICES=1 python train.py \
    --agent edward --wandb_name "edward/cond-interp-s42" --wandb_group condition-interp-augmentation \
    --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
    --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
    --pressure_first --pressure_deep --residual_prediction --surface_refine \
    --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
    --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
    --condition_interp --condition_interp_weight 0.3
  ```

### Seed 73 — W&B run `65eo5fnu`
- **Best epoch:** 137 (wall-clock timeout at 181.1 min)
- **val/loss:** 0.6352
- **Surface MAE (p):**
  - p_in: 14.3 (baseline 11.709, **+22.1% worse**)
  - p_tan: 28.2 (baseline 27.402, **+2.9% worse**)
  - p_oodc: 8.0 (baseline 7.544, **+6.0% worse**)
  - p_re: 7.0 (baseline 6.481, **+8.0% worse**)
- **Peak VRAM:** 67.8 GB

### Summary

| Metric | Baseline | Seed 42 | Seed 73 | Δ (avg) |
|--------|----------|---------|---------|---------|
| p_in   | 11.709   | 13.3    | 14.3    | **+17.9% worse** |
| p_tan  | 27.402   | 27.5    | 28.2    | **+1.7% worse** |
| p_oodc | 7.544    | 7.8     | 8.0     | **+4.7% worse** |
| p_re   | 6.481    | 6.9     | 7.0     | **+7.3% worse** |
| val/loss | ~0.37  | 0.6246  | 0.6352  | significantly worse |

### What happened

**The condition-space interpolation augmentation hurt substantially across all metrics.** The val/loss of ~0.63 vs ~0.37 baseline confirms this is a significant regression, not noise.

**Why it didn't work:**

1. **VRAM overhead degraded effective training**. Appending 4 interp samples per batch expanded it ~25%, consuming 67GB vs ~40GB baseline. More critically, the larger effective batch size changed the gradient dynamics — the model trained fewer effective steps for the same wall-clock time (slower epochs at ~80s vs ~67s).

2. **The DSDF cosine similarity proxy finds too many spurious pairs.** Checking `x[:, 0, 2:10]` (first node's 8-dim DSDF) is not a reliable geometry identifier. Different geometries can have similar first-node DSDF values, causing genuinely different-geometry pairs to be interpolated. This introduces physically invalid interpolated labels — exactly the failure mode that sank the cross-geometry mixup (#2327).

3. **The interpolated labels create conflicting supervision.** The model is simultaneously trained on real CFD data and approximate pseudo-solutions. For the RANS problem, the approximation error in interpolated labels appears to be large enough to actively harm learning, not just add noise.

4. **Condition interpolation weight of 0.3 is too high.** With 4 interp samples at weight 0.3, they contribute ~25% of batch weight — a significant fraction of the total loss signal. A much smaller weight (e.g., 0.05-0.1) might be needed if this approach is revisited.

### Suggested follow-ups
- If revisiting: use a proper geometry ID from the dataset (e.g., NACA code or explicit shape parameters) rather than DSDF cosine similarity for pair matching.
- Much lower interp weight (0.05) to reduce the impact of label noise.
- Cap interp samples at 2 per batch (not 4) to reduce VRAM overhead.
- Consider condition interpolation for OOD-Re only (Re-direction is safer than AoA for RANS linearity, as boundary layer thickness scales as Re^{-1/5}).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2345 because ADVISOR: Closing — condition-space interpolation caused significant regression across all metrics (p_in +17.9%, p_oodc +4.7%, p_tan +1.7%, p_re +7.3%). Three compounding issues: (1) DSDF cosine similarity is an unreliable geometry proxy — different geometries can match, creating physically invalid interpolated labels; (2) 67GB VRAM (vs 40GB baseline) from appended interp samples slowed training; (3) interp weight 0.3 is far too high for approximate pseudo-labels. Good analysis of why it failed. The core lesson: same-geometry pair matching requires an explicit geometry ID, not feature similarity. Direction exhausted for now..
